File: leetcode-submissions/547_Number_of_Provinces.ts
class DSU{
    parent:number[];
    rank:number[];

    constructor(size:number){
        this.parent=new Array(size).fill(0).map((_,index)=>index);
        this.rank=new Array(size).fill(0);
    }
    find(x:number):number{
    return this.parent[x]===x?x:this.find(this.parent[x]);
    }
    union(u:number ,v:number):void{
        let parentU:number =this.find(u);
        let parentV:number =this.find(v);
        if(parentU===parentV){return ;}
        if(parentU>=parentV){
            this.parent[parentV]=parentU;
            this.rank[parentU]++;
        }else{
            this.parent[parentU]=parentV;
            this.rank[parentV]++;
        }

    }
}
function findCircleNum(isConnected: number[][]): number {
    const dsu=new DSU(isConnected.length);
    for(let i:number=0;i<isConnected.length;i++){
        for(let j:number =0;j<isConnected[0].length;j++){
            if(isConnected[i][j]===1){
                dsu.union(i,j);
            }
        }
    }
    let cc:number =0;
    for(let i:number =0;i<isConnected.length;i++){
        if(dsu.find(i)===i){
            cc++;
        }
    }
    return cc;
};

File: leetcode-submissions/547_Number_of_Provinces.cpp
class Solution {
public:
    int find(vector<int>&parent,int x){
        return parent[x]=(parent[x]==x)?x:find(parent,parent[x]);
    }
    void Union(vector<int>&parent,vector<int>&rank,int a,int b){
        a=find(parent,a);
        b=find(parent,b);
        if(a==b){
            return ;
        }
        if(rank[a]>=rank[b]){
            rank[a]++;
            parent[b]=a;
        }else{
            rank[b]++;
            parent[a]=b;
        }
    }
    
    int findCircleNum(vector<vector<int>>& isConnected) {
        //now finding the solution of this question by using the union find data structure in whihc we need to calculate the number of connected components only
        //in this question we are given 1 based indexing
        int n=isConnected.size();
        vector<int>parent(n+1);
        vector<int>rank(n+1,0);
        for(int i=1;i<=n;i++){
            parent[i]=i;
        }
        for(int i=0;i<n;i++){
            for(int j=0;j<isConnected[i].size();j++){
                if(isConnected[i][j]==1 ){

                Union(parent,rank,i+1,j+1);
                }
            }
        }
        int cnt=0;
        for(int i=1;i<=n;i++){
            if(parent[i]==i){
                cnt++;
            }
        }
        return cnt;

    }
};